File: tools/correlator/src/capture.rs
use std::io::{self, Write};
use std::path::Path;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use rusqlite::params;

use crate::constants;
use crate::db::{init_db, default_db_path};
use crate::context::parse_nmap_xml;
use crate::tools::sudo_cmd;
use crate::tshark::{extract_fields, tshark_args};

pub fn run_capture(interface: &str, target: &str, duration: u64, no_save: bool, output: Option<&Path>,
                   fast: bool, no_nmap: bool, no_tshark: bool, debug: bool, stealth_level: u8) {
    let db_path = default_db_path(no_save, output);
    println!("═══════ CAPTURE ═══════");
    println!("[System] Database: {}", db_path.display());
    println!("[System] Stealth level: {} ({})", stealth_level, match stealth_level {
        0 => "full scan — aggressive nmap, background scanner active",
        1 => "light — rate-limited scan, slower background scanner",
        _ => "passive — TShark only, no active scanning",
    });

    let conn = init_db(&db_path);
    let conn = Arc::new(Mutex::new(conn));

    let tshark_child = if !no_tshark {
        println!("[System] Starting TShark on {} for {}s...", interface, duration);
        let args = tshark_args(interface, "");
        let child = sudo_cmd("tshark").args(&args)
            .stdin(std::process::Stdio::inherit())
            .stdout(std::process::Stdio::piped()).stderr(std::process::Stdio::null())
            .spawn().expect("Failed to start tshark");
        Some(child)
    } else {
        None
    };

    if !no_nmap && stealth_level <= constants::STEALTH_LIGHT {
        println!("[System] Starting nmap scan of {}...", target);
        let mut args: Vec<&str> = constants::nmap_flags(stealth_level, fast).to_vec();
        args.push(target);

        let output = sudo_cmd("nmap").args(&args)
            .stdin(std::process::Stdio::inherit())
            .output().expect("Failed to run nmap");

        let xml_str = String::from_utf8_lossy(&output.stdout);
        if !xml_str.is_empty() {
            let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs_f64();
            let c = conn.lock().unwrap();
            let summary = parse_nmap_xml(&xml_str, &c, now);
            c.execute(
                "INSERT INTO nmap_scans (target, scan_time, raw_xml, summary) VALUES (?1, ?2, ?3, ?4)",
                params![target, now, xml_str.to_string(), summary.clone()],
            ).expect("Failed to store scan");
            let device_count: u64 = c.query_row("SELECT COUNT(*) FROM devices", [], |r| r.get(0)).unwrap_or(0);
            println!("[nmap] Found {} devices:\n{}", device_count, summary);
        }
    }

    if let Some(mut child) = tshark_child {
        let child_pid = child.id();
        let timer_handle = std::thread::spawn(move || {
            std::thread::sleep(Duration::from_secs(duration));
            let _ = sudo_cmd("kill").args(["-INT", &child_pid.to_string()]).output();
        });

        if let Some(stdout_stream) = child.stdout.take() {
            let reader = std::io::BufReader::new(stdout_stream);
            let mut packet_count: u64 = 0;
            let mut stored_count: u64 = 0;
            let start = Instant::now();

            use std::io::BufRead;

            for line_result in reader.lines() {
                match line_result {
                    Ok(raw_line) => {
                        if raw_line.trim().is_empty() { continue; }
                        if raw_line.contains("\"index\"") && !raw_line.contains("\"_source\"") { continue; }

                        if debug {
                            eprintln!("[debug] {}", raw_line.trim());
                        }

                        if let Some(pkt) = extract_fields(&raw_line) {
                            packet_count += 1;
                            let c = conn.lock().unwrap();
                            let _ = c.execute(
                                "INSERT INTO packets (epoch, ip_src, ip_dst, tcp_src_port, tcp_dst_port, udp_src_port, udp_dst_port, dns_query, raw_json, frame_len) VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9,?10)",
                                params![pkt.epoch, pkt.ip_src, pkt.ip_dst, pkt.tcp_src_port.map(|p| p as i32), pkt.tcp_dst_port.map(|p| p as i32), pkt.udp_src_port.map(|p| p as i32), pkt.udp_dst_port.map(|p| p as i32), pkt.dns_query, raw_line.trim(), pkt.frame_len.map(|p| p as i32)]
                            );
                            stored_count += 1;

                            if packet_count.is_multiple_of(100) {
                                let elapsed = start.elapsed().as_secs();
                                print!("\r\x1b[K  {} captured, {} stored, {}s elapsed  ", packet_count, stored_count, elapsed);
                                io::stdout().flush().ok();
                            }
                        }
                    }
                    Err(_) => break,
                }
            }

            let _ = child.kill();
            let _ = child.wait();
            let _ = timer_handle.join();

            println!("\n\n═══ CAPTURE COMPLETE ═══");
            println!("[System] {} packets captured, {} stored", packet_count, stored_count);
        }
    }

    let device_count: u64 = conn.lock().unwrap().query_row("SELECT COUNT(*) FROM devices", [], |r| r.get(0)).unwrap_or(0);
    println!("[System] {} devices in database", device_count);
    println!("[System] Database saved at {}", db_path.display());
    println!("[System] To chat: correlator chat -d {}", db_path.display());
}

File: internal/capture/capture.go
// Package capture orchestrates parallel TShark packet capture and nmap host
// discovery into a SQLite database.
package capture

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"os/exec"
	"strings"
	"sync"
	"time"

	"github.com/hxmbl/hx_netkit/internal/nmap"
	"github.com/hxmbl/hx_netkit/internal/store"
	"github.com/hxmbl/hx_netkit/internal/tshark"
)

// Options configures one capture run.
type Options struct {
	Interface    string
	Target       string
	DurationSecs uint64
	Fast         bool
	NoNmap       bool
	NoTShark     bool
	Debug        bool
	StealthLevel uint8
	DBPath       string
	Out          io.Writer // progress output; os.Stdout by default
}

func sudoCmd(prog string, args ...string) *exec.Cmd {
	isRoot := false
	if out, err := exec.Command("id", "-u").Output(); err == nil {
		isRoot = strings.TrimSpace(string(out)) == "0"
	}
	if isRoot {
		return exec.Command(prog, args...)
	}
	full := append([]string{prog}, args...)
	return exec.Command("sudo", full...)
}

// Run executes the capture and returns (packetsStored, devicesFound, error).
func Run(opts Options) (int, int, error) {
	out := opts.Out
	if out == nil {
		out = os.Stdout
	}

	fmt.Fprintln(out, "═══════ CAPTURE ═══════")
	fmt.Fprintf(out, "[System] Database: %s\n", opts.DBPath)
	level := "passive — TShark only, no active scanning"
	switch opts.StealthLevel {
	case nmap.StealthFull:
		level = "full scan — aggressive nmap, background scanner active"
	case nmap.StealthLight:
		level = "light — rate-limited scan, slower background scanner"
	}
	fmt.Fprintf(out, "[System] Stealth level: %d (%s)\n", opts.StealthLevel, level)

	db, err := store.Open(opts.DBPath)
	if err != nil {
		return 0, 0, err
	}
	defer db.Close()

	var wg sync.WaitGroup
	stored := 0

	// ── nmap side ──
	if !opts.NoNmap && opts.StealthLevel <= nmap.StealthLight {
		wg.Add(1)
		go func() {
			defer wg.Done()
			flags := nmap.Flags(opts.StealthLevel, opts.Fast, opts.Target)
			if flags == nil {
				return
			}
			fmt.Fprintf(out, "[System] Starting nmap scan of %s...\n", opts.Target)
			cmd := sudoCmd("nmap", flags...)
			var stderrBuf strings.Builder
			cmd.Stderr = &stderrBuf
			xmlOut, err := cmd.Output()
			if err != nil {
				msg := err.Error()
				if tail := strings.TrimSpace(stderrBuf.String()); tail != "" {
					msg += ": " + tail
				}
				fmt.Fprintf(out, "[Error] Failed to run nmap: %s. Is it installed?\n", msg)
				return
			}
			if len(xmlOut) == 0 {
				return
			}
			now := float64(time.Now().UnixNano()) / 1e9
			devices := nmap.ParseXML(xmlOut)
			for _, d := range devices {
				db.UpsertDevice(d.IP, d.MAC, d.Hostname, d.MACVendor, d.OSGuess, d.PortsString(), now)
			}
			summary := nmap.Summarize(devices)
			db.RecordScan(opts.Target, string(xmlOut), summary, now)
			fmt.Fprintf(out, "[nmap] Found %d devices:\n%s\n", len(devices), summary)
		}()
	} else {
		fmt.Fprintln(out, "[System] Skipping nmap (disabled or passive stealth).")
	}

	// ── tshark side ──
	if !opts.NoTShark {
		fmt.Fprintf(out, "[System] Starting TShark on %s for %ds...\n", opts.Interface, opts.DurationSecs)
		stored, err = runTShark(db, opts, out)
		if err != nil {
			wg.Wait()
			return stored, countDevices(db), err
		}
	}

	wg.Wait()

	devices := countDevices(db)
	fmt.Fprintf(out, "\n\n═══ CAPTURE COMPLETE ═══\n")
	fmt.Fprintf(out, "[System] %d packets captured, %d stored\n", stored, stored)
	fmt.Fprintf(out, "[System] %d devices in database\n", devices)
	fmt.Fprintf(out, "[System] Database saved at %s\n", opts.DBPath)
	fmt.Fprintf(out, "[System] To chat: correlator chat -d %s\n", opts.DBPath)
	return stored, devices, nil
}

func countDevices(db *store.DB) int {
	s, err := db.Stats()
	if err != nil {
		return 0
	}
	return int(s.Devices)
}

func runTShark(db *store.DB, opts Options, out io.Writer) (int, error) {
	cmd := sudoCmd("tshark", tshark.Args(opts.Interface, "")...)
	stdoutPipe, err := cmd.StdoutPipe()
	if err != nil {
		return 0, fmt.Errorf("failed to pipe tshark stdout: %w", err)
	}
	var stderrBuf strings.Builder
	cmd.Stderr = &stderrBuf
	if err := cmd.Start(); err != nil {
		msg := err.Error()
		if tail := strings.TrimSpace(stderrBuf.String()); tail != "" {
			msg += ": " + tail
		}
		return 0, fmt.Errorf("failed to start tshark: %s. Is it installed?", msg)
	}

	// Timer can be cancelled early when the stream ends on its own, so the
	// goroutine never outlives the capture.
	timerStop := make(chan struct{})
	var timerOnce sync.Once
	defer timerOnce.Do(func() { close(timerStop) })
	go func() {
		select {
		case <-time.After(time.Duration(opts.DurationSecs) * time.Second):
			interrupt(cmd)
		case <-timerStop:
		}
	}()

	count := 0
	scanner := bufio.NewScanner(stdoutPipe)
	scanner.Buffer(make([]byte, 0, 256*1024), 1024*1024)
	start := time.Now()
	for scanner.Scan() {
		rawLine := scanner.Text()
		if tshark.Skippable(rawLine) {
			continue
		}
		pkt, ok := tshark.ParseLine(rawLine)
		if !ok {
			continue
		}
		count++
		_ = db.InsertPacket(
			pkt.Epoch, pkt.IPSrc, pkt.IPDst,
			i64(pkt.TCPsrc), i64(pkt.TCPdst), i64(pkt.UDPsrc), i64(pkt.UDPdst),
			pkt.DNSQuery, strings.TrimSpace(rawLine), i64(pkt.FrameLen),
		)
		if opts.Debug {
			fmt.Fprintf(os.Stderr, "[debug] %s\n", rawLine)
		}
		if count%100 == 0 {
			elapsed := int(time.Since(start).Seconds())
			fmt.Fprintf(out, "\r\x1b[K  %d captured, %d stored, %ds elapsed  ", count, count, elapsed)
		}
	}
	kill(cmd)
	wait(cmd)

	if count == 0 {
		if tail := strings.TrimSpace(stderrBuf.String()); tail != "" {
			fmt.Fprintf(out, "\n[Warn] tshark produced no packets. stderr: %s\n", tail)
		}
	}

	return count, nil
}

func interrupt(cmd *exec.Cmd) {
	if cmd.Process != nil {
		_ = cmd.Process.Signal(os.Interrupt)
	}
}

// SudoTShark starts tshark with privilege escalation when needed.
func SudoTShark(args ...string) *exec.Cmd { return sudoCmd("tshark", args...) }

// Interrupt sends SIGINT to a running process.
func Interrupt(cmd *exec.Cmd) { interrupt(cmd) }

// Kill force-terminates a running process.
func Kill(cmd *exec.Cmd) { kill(cmd) }

// Wait reaps a finished child process.
func Wait(cmd *exec.Cmd) { wait(cmd) }

// StreamLines reads r line by line, invoking fn per line while fn returns true.
func StreamLines(r io.Reader, fn func(string) bool) error {
	scanner := bufio.NewScanner(r)
	scanner.Buffer(make([]byte, 0, 256*1024), 1024*1024)
	for scanner.Scan() {
		if !fn(scanner.Text()) {
			break
		}
	}
	return scanner.Err()
}

func kill(cmd *exec.Cmd) {
	if cmd.Process != nil {
		_ = cmd.Process.Kill()
	}
}

func wait(cmd *exec.Cmd) {
	if cmd.Process != nil {
		_, _ = cmd.Process.Wait()
	}
}

func i64(v uint32) int64 { return int64(v) }
